File: leetcode-java/src/main/java/chapter02/CircleList.java
package chapter02;

import utils.ListNode;
import java.util.Set;
import java.util.HashSet;

/**
 * <h2>环形链表</h2>
 * <h3>1. 环形链表</h3>
 * <h3>2. 环形链表 II</h3>
 * <h3>注: 这两个题思路基本没有区别</h3>
 */
public class CircleList {

    /**
     * <h3>思路: 环形链表 II</h3>
     */
    private static ListNode detectCycleHashSet(ListNode head){
        Set<ListNode> set = new HashSet<>();
        ListNode current = head;
        while (current != null){
            if (!set.add(current))
                return current;
            current = current.next;
        }
        return null;
    }

    /**
     * <h3>思路: 环形链表 II</h3>
     */
    public static ListNode detectCycle(ListNode head){
        ListNode slow = head, fast = head;
        do {
           if (fast == null || fast.next == null) return null;
           fast = fast.next.next;
           slow = slow.next;
        }while (slow != fast);
        fast = head;
        while (fast != slow){
            slow = slow.next;
            fast = fast.next;
        }
        return fast;
    }
}

File: leetcode-go/leetcode/chapter02/circlelist.go
package chapter02

// 环形链表: 哈希表
func hasCycle1(head *ListNode) *ListNode {
	current := head
	set := make(map[*ListNode]string)
	for current != nil {
		if _, ok := set[current]; ok {
			return current
		}
		set[current] = ""
		current = current.Next
	}
	return nil
}

// 环形链表: 快慢指针
func hasCycle2(head *ListNode) *ListNode {
	flag := true
	slow, fast := head, head

	for flag || slow != fast {
		if fast == nil || fast.Next == nil {
			return nil
		}
		flag = false
		slow = slow.Next
		fast = fast.Next
	}
	fast = head
	for slow != fast {
		slow = slow.Next
		fast = fast.Next
	}
	return slow
}
